statics/js/components/graph-layout.js

```diff
@@ -1,6 +1,184 @@
+
+var LinkLabelBandwidth = Vue.extend({
+  mixins: [apiMixin],
+
+  methods: {
+
+    bandwidthFromMetrics: function(metrics) {
+      if (!metrics) {
+        return 0;
+      }
+      if (!metrics.Last) {
+        return 0;
+      }
+      if (!metrics.Start) {
+        return 0;
+      }
+
+      const totalByte = (metrics.RxBytes || 0) + (metrics.TxBytes || 0);
+      const deltaMillis = metrics.Last - metrics.Start;
+      const elapsedMillis = Date.now() - new Date(metrics.Last);
+
+      if (deltaMillis === 0) {
+        return 0;
+      }
+      if (elapsedMillis > maxClockSkewMillis) {
+        return 0;
+      }
+
+      return Math.floor(8 * totalByte * 1000 / deltaMillis); // bits-per-second 
+    },
+
+    setup: function(topology) {
+      this.topology = topology;
+    },
+
+    updateData: function(link) {
+      if (!link.target.metadata.LastUpdateMetric) {
+        return;
+      }
+
+      const defaultBandwidthBaseline = 1024 * 1024 * 1024; // 1 gbps
+
+      link.bandwidthBaseline = (this.topology.bandwidth.bandwidthThreshold === 'relative') ?
+        link.target.metadata.Speed || defaultBandwidthBaseline : 1;
+
+      link.bandwidthAbsolute = this.bandwidthFromMetrics(link.target.metadata.LastUpdateMetric);
+      link.bandwidth = link.bandwidthAbsolute / link.bandwidthBaseline;
+    },
+ 
+    hasData: function(link) {
+      if (!link.target.metadata.LastUpdateMetric) {
+        return false;
+      }
+
+      if (!link.bandwidth) {
+        return false;
+      }
+      return link.bandwidth > this.topology.bandwidth.active;
+    },
+
+    getText: function(link) {
+      return bandwidthToString(link.bandwidthAbsolute);
+    },
+
+    isActive: function(link) {
+      return (link.bandwidth > this.topology.bandwidth.active) && (link.bandwidth < this.topology.bandwidth.warning);
+    },
+
+    isWarning: function(link) {
+      return (link.bandwidth >= this.topology.bandwidth.warning) && (link.bandwidth < this.topology.bandwidth.alert);
+    },
+
+    isAlert: function(link) {
+      return link.bandwidth >= this.topology.bandwidth.alert;
+    },
+  },
+});
+
+var LinkLabelLatency = Vue.extend({
+  mixins: [apiMixin],
+
+  methods: {
+    setup: function(topology) {
+      this.active = 0;
+      this.warning = 10
+      this.alert = 100;
+    },
+
+    flowKey: function(a) {
+      return `TrackingID:${a.TrackingID},ANodeTID:${a.ANodeTID},BNodeTID:${a.BNodeTID}`;
+    },
+
+    flowCompare: function(a, b) {
+      const keya = this.flowKey(a);
+      const keyb = this.flowKey(b);
+      return keya.localeCompare(keyb);
+    },
+
+    updateLatency: function(link, flows) {
+      let prev, curr;
+      for (let i in flows) {
+        prev = curr;
+        curr = flows[i];
+
+        if (!prev) {
+          continue;
+        }
+
+        if (this.flowKey(curr) != this.flowKey(prev)) {
+          continue;
+        }
+
+        link.latencyTimestamp = Math.max(curr.Last, prev.Last);
```

Comment:
you should use Date.Now() as comparing packet Last (capture) clock with Now (local browser clock) is not correct

Reply:
If the flow we based our calculations ran 1 hour ago or 1 day ago (as would be apparent from `curr.Last` and `prev.Last`) do we still want to use it? No! So to mitigate this I *do* want to check the time recorded during capture and if flow is within a set window (using 5 minutes skew time as used for bandwidth calculation) - then use it.

The 5 minute window was intentionally taken so as to also mitigate time skew between the agent and the browser host.